Barbara needs async helpers (#121)
rust-lang/wg-async-foundations: src/vision/status_quo/barbara_needs_async_helpers.md

  --> src/main.rs:11:17
   |
11 |             n + sum(n - 1).await
   |                 ^^^^^^^^^^^^^^^^ the trait `Unpin` is not implemented for `dyn Future<Output = usize>`
   |
   = note: required because of the requirements on the impl of `Future` for `Box<dyn Future<Output = usize>>`
   = note: required by `poll`
```

She then reads about `Unpin` and `Pin`, and finally comes up with a solution.

```rust
fn sum(n: usize) -> Pin<Box<dyn Future<Output = usize>>> {
    Box::pin(async move {
        if n == 0 {
            0
        } else {
            n + sum(n - 1).await
        }
    })
}
```

The code works!

She searches online for better methods and finds out the [async-book](https://rust-lang.github.io/async-book/01_getting_started/01_chapter.html). She reads about [recursion](https://rust-lang.github.io/async-book/07_workarounds/04_recursion.html) and finds out a cleaner way using the [`futures`] crate.

```rust
use futures::future::{BoxFuture, FutureExt};

fn sum(n: usize) -> BoxFuture<'static, usize> {
    async move {
        if n == 0 {
            0
        } else {
            n + sum(n - 1).await
        }
    }.boxed()
}
```

> She also asks one of her peers for a code review asynchronously, and after awaiting their response, she learns about the [`async-recursion`] crate. Then she adds [`async-recursion`] to the dependencies.

Reviewer: I guess I expected one last example of how it works with the crate. 

```suggestion
She also asks one of her peers for a code review asynchronously, and after awaiting their response, she learns about the [`async-recursion`] crate. Then she adds [`async-recursion`] to the dependencies and writes the following:

#[async_recursion]
async fn sum(n: usize) -> usize {
        if n == 0 {
            0
        } else {
            n + sum(n - 1).await
        }
}

```

If you would like to expand on this story, or adjust the answers to the FAQ, feel free to open a PR making edits (but keep in mind that, as they reflect peoples' experiences, status quo stories [cannot be wrong], only inaccurate). Alternatively, you may wish to [add your own status quo story][htvsq]!

## The story

[Barbara], an experienced Rust user, is prototyping an async Rust service for work. To get things working quickly, she decides to prototype in tokio, since it is unclear which runtime her work will use.

She starts adding warp and tokio to her dependencies list. She notices that warp suggests using tokio with the `full` feature. She's a bit concerned about how this might affect the compile times and also that *all* of tokio is needed for her little project, but she pushes forward.

As she builds out functionality, she's pleased to see tokio provides a bunch of helpers like `join!` and async versions of the standard library types like channels and mutexes.

> After completing one endpoint, she moves to a new one which requires streaming http responses to the client. Barbara quickly finds out that tokio does not provide a stream type, and so she adds `tokio-stream` to her dependencies.

Reviewer: See my comment below for more details about this...

> After completing one endpoint, she moves to a new one which requires streaming http responses to the client. Barbara quickly finds out that tokio does not provide a stream type, and so she adds `tokio-stream` to her dependencies.

Author: She might have learned about `tokio-stream` from `tokio` docs [itself](https://docs.rs/tokio/1.4.0/tokio/stream/index.html). Is this ok?

> After completing one endpoint, she moves to a new one which requires streaming http responses to the client. Barbara quickly finds out that tokio does not provide a stream type, and so she adds `tokio-stream` to her dependencies.

Reviewer: sure!

> Moving on she tries to make some functions generic over the web framework underneath, so she tries to abstract off the functionality to a trait. But Rust doesn't support async functions in traits yet, so she adds `async_trait` to her dependencies.

Author: Something like this:

---

She writes an async function inside a trait, just like a normal function.

```rust=
trait Example {
    async fn foo();
}
```

Then she gets a helpful error message.

```=
error[E0706]: functions in traits cannot be declared `async`
 --> src/lib.rs:2:5
  |
2 |     async fn foo();
  |     -----^^^^^^^^^^
  |     |
  |     `async` because of this
  |
  = note: `async` trait functions are not currently supported
  = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait
```

---

I'll add this, and also the link to the crate.

> Moving on she tries to make some functions generic over the web framework underneath, so she tries to abstract off the functionality to a trait. But Rust doesn't support async functions in traits yet, so she adds `async_trait` to her dependencies.

Reviewer: thanks :)

> Some of her functions are recursive, so to make them async she starts boxing her futures. Then she learns about `async-recursion` and then adds it to the dependencies.

Author: Yeah that could happen, or else, she google's "recursive async function in rust" it and finds out the async-recursion crate documentation? Also, I'll add links to all the crates.

> Some of her functions are recursive, so to make them async she starts boxing her futures. Then she learns about `async-recursion` and then adds it to the dependencies.

Reviewer: I'd like to have her try it out by hand first just to show what that is like

> Her stream implementation needed `pin_project` so she brings that also as a dependency.

Reviewer 2: I’m not sure off the top of my head. It should roughly be the same limitations as async fn with future. The returned type is anonymous and it wouldn’t work with “resource” types.

> "Finally!", Barbara says, breathing a sigh of relief. She is done with her prototype, and shows it off at work, but to her dismay, the team decides that tokio is not an appropriate runtime for their use case. Barbara's heart skips a beat. "Oh no, what to do now," she thinks.

Reviewer: ```suggestion
"Finally!", Barbara says, breathing a sigh of relief. She is done with her prototype, and shows it off at work, but to her dismay, the team decides that they need to use a custom runtime for their use case. They're building an embedded system and it has relatively limited resources. Barbara thinks, "No problem, it should be easy enough to change runtimes, right?"
```

Reviewer: Maybe she can be hopeful :)

So now Barbara starts the journey of replacing tokio with a myriad of off the shelf and custom helpers. She can't use warp so now she has to find an alternative. She also has to find a new channel implementations and there are a few:
* In `futures`
* `async-std` has one, but it seems to be tied to another runtime so she can't use that.
* `smol` has one that is independent.

> This process of "figure out which alternative is an option" is repeated many times. She also tries to use the `select!` macro from `futures` but it requires more pinning and workarounds (and then her stack overflows).

Reviewer: ```suggestion
This process of "figure out which alternative is an option" is repeated many times. She also tries to use the `select!` macro from `futures` but it requires more pinning and workarounds (not to mention a [stack overflow](https://rust-lang.github.io/wg-async-foundations/vision/status_quo/alan_runs_into_stack_trouble.html) or two).
```

```rust
trait Client {
    async fn get();
}
```

Then she gets a helpful error message.

```
error[E0706]: functions in traits cannot be declared `async`
 --> src/lib.rs:2:5
  |
2 |     async fn get();
  |     -----^^^^^^^^^^
  |     |
  |     `async` because of this
  |
  = note: `async` trait functions are not currently supported
  = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait
```

She then realizes that Rust doesn't support async functions in traits yet, so she adds [`async-trait`] to her dependencies.

Some of her functions are recursive, and she wanted them to be async functions, so she sprinkles some `async/.await` keywords in those functions.

```rust
async fn sum(n: usize) -> usize {
    if n == 0 {
        0
    } else {
        n + sum(n - 1).await
    }
}
```

Then she gets an error message.

```
error[E0733]: recursion in an `async fn` requires boxing
 --> src/lib.rs:1:27
  |
1 | async fn sum(n: usize) -> usize {
  |                           ^^^^^ recursive `async fn`
  |
  = note: a recursive `async fn` must be rewritten to return a boxed `dyn Future`
```

> So to make these functions async she starts boxing her futures. She also asks one of her peers for a code review asynchronously, and after awaiting their response, she learns about the [`async-recursion`] crate. Then she adds [`async-recursion`] to the dependencies.

Reviewer: > So to make these functions async she starts boxing her futures.

It'd be nice to see a code example of what this looks like

Reviewer: @rylev I think you said you were boxing your futures like a barbarian, maybe you can provide an example?

Author: > > So to make these functions async she starts boxing her futures.
> 
> It'd be nice to see a code example of what this looks like

What about something like this?

---

She tries boxing her futures, the hard way, fighting with the compiler. She knows that `async` keyword is sort of a sugar for `impl Future` so she tries the following at first.

```rust
fn sum(n: usize) -> Box<dyn Future<Output = usize>> {
    Box::new(async move {
        if n == 0 {
            0
        } else {
            n + sum(n - 1).await
        }
    })
}
```

The compiler gives an error.

```
error[E0277]: `dyn Future<Output = usize>` cannot be unpinned
  --> src/main.rs:11:17
   |
11 |             n + sum(n - 1).await
   |                 ^^^^^^^^^^^^^^^^ the trait `Unpin` is not implemented for `dyn Future<Output = usize>`
   |
   = note: required because of the requirements on the impl of `Future` for `Box<dyn Future<Output = usize>>`
   = note: required by `poll`
```

She then reads about `Unpin` and `Pin`, and then finally comes up with a solution.

```rust
fn sum(n: usize) -> Pin<Box<dyn Future<Output = usize>>> {
    Box::pin(async move {
        if n == 0 {
            0
        } else {
            n + sum(n - 1).await
        }
    })
}
```

The code works!

---

I assumed here that she would take a route without involving the `futures` crate. With the `futures` crate, it is as easy as:

```rust
use futures::future::{BoxFuture, FutureExt};

fn sum(n: usize) -> BoxFuture<'static, usize> {
    async move {
        if n == 0 {
            0
        } else {
            n + sum(n - 1).await
        }
    }.boxed()
}
```

Reviewer: that looks great, thanks. As I expected, it's a significant change -- she has to move from `async fn` to a `fn`. In this case, the signature is fairly easy, but it can be quite a bit more difficult if regions are involved.

Reviewer: I would probably have her write it by hand then later learn about futures crate, then about async-recursion.

Reviewer 3: This looks great!

Author: > that looks great, thanks. As I expected, it's a significant change -- she has to move from `async fn` to a `fn`. In this case, the signature is fairly easy, but it can be quite a bit more difficult if lifetimes are involved.

Is an example with lifetimes also required?

Reviewer: Nah, I don't think that's necessary. You could add a FAQ about it though.

> As she is working, she realizes that what she really needs is to write a `Stream` of data. She starts trying to write her `Stream` implementation and spends several hours banging her head against her desk in frustration (her challenges are pretty similar to what [Alan experienced](https://rust-lang.github.io/wg-async-foundations/vision/status_quo/alan_hates_writing_a_stream.html)). Ultimately she's stuck trying to figure out why her `&mut self.foo` call is giving her errors:

Reviewer: ```suggestion
As she is working, she realizes that what she really needs is to write a `Stream` of data. She starts trying to write her `Stream` implementation and spends several hours banging her head against her desk in frustration (her challenges are pretty similar to what [Alan experienced](./alan_hates_writing_a_stream.md)). Ultimately she's stuck trying to figure out why her `&mut self.foo` call is giving her errors:
```
